cocotb/share/lib/vpi/VpiImpl.cpp

```diff
@@ -296,16 +296,36 @@ GpiObjHdl* VpiImpl::native_check_create(void *raw_hdl, GpiObjHdl *parent)
 GpiObjHdl* VpiImpl::native_check_create(std::string &name, GpiObjHdl *parent)
 {
     vpiHandle new_hdl;
+    vpiHandle parent_hdl = parent->get_handle<vpiHandle>();
     std::string fq_name = parent->get_fullname() + "." + name;
     std::vector<char> writable(fq_name.begin(), fq_name.end());
     writable.push_back('\0');
 
     new_hdl = vpi_handle_by_name(&writable[0], NULL);
 
-    /* No need to iterate to look for generate loops as the tools will at least find vpiGenScopeArray */
+    /* If not found, check to see if the name of a generate loop.
+     * Icarus does not support vpiGenScopeArray, only vpiGenScope.
+     */
     if (new_hdl == NULL) {
-        LOG_DEBUG("Unable to query vpi_get_handle_by_name %s", fq_name.c_str());
-        return NULL;
+        vpiHandle iter = vpi_iterate(vpiInternalScope, parent_hdl);
+
+        if (iter != NULL) {
+            vpiHandle rgn;
+            for (rgn = vpi_scan(iter); rgn != NULL; rgn = vpi_scan(iter)) {
+                if (vpi_get(vpiType, rgn) == vpiGenScope) {
+                    std::string rgn_name = vpi_get_str(vpiName, rgn);
+                    if (rgn_name.compare(0,name.length(),name) == 0) {
+                        new_hdl = parent_hdl;
+                        vpi_free_object(iter);
+                        break;
+                    }
+                }
+            }
+        }
```

Comment:
I might refactor this slightly doing a `iter == NULL` check after the call to `vpi_iterate`, logging failure and returning. Then you can de-indent this loop 1 level.